services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/handlers/CopySourceInterceptor.java

```diff
@@ -0,0 +1,127 @@
+ * <p>
+ * <a href="https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html#API_UploadPartCopy_RequestParameters">API_UploadPartCopy_RequestParameters</a>
+ */
+@SdkInternalApi
+public final class CopySourceInterceptor implements ExecutionInterceptor {
+
+    @Override
+    public SdkRequest modifyRequest(ModifyRequest context, ExecutionAttributes executionAttributes) {
+        SdkRequest request = context.request();
+        if (request instanceof CopyObjectRequest) {
+            return modifyCopyObjectRequest((CopyObjectRequest) request);
+        }
+        if (request instanceof UploadPartCopyRequest) {
+            return modifyUploadPartCopyRequest((UploadPartCopyRequest) request);
+        }
+        return request;
+    }
+
+    private static SdkRequest modifyCopyObjectRequest(CopyObjectRequest request) {
+        if (request.copySource() != null) {
+            requireNotSet(request.sourceBucket(), "sourceBucket");
+            requireNotSet(request.sourceKey(), "sourceKey");
+            requireNotSet(request.sourceVersionId(), "sourceVersionId");
+            return request;
+        }
+        String copySource = constructCopySource(
+            requireSet(request.sourceBucket(), "sourceBucket"),
+            requireSet(request.sourceKey(), "sourceKey"),
+            request.sourceVersionId()
+        );
+        return request.toBuilder()
+                      .sourceBucket(null)
+                      .sourceKey(null)
+                      .sourceVersionId(null)
+                      .copySource(copySource)
+                      .build();
+    }
+
+    private static SdkRequest modifyUploadPartCopyRequest(UploadPartCopyRequest request) {
+        if (request.copySource() != null) {
+            requireNotSet(request.sourceBucket(), "sourceBucket");
+            requireNotSet(request.sourceKey(), "sourceKey");
+            requireNotSet(request.sourceVersionId(), "sourceVersionId");
+            return request;
+        }
+        String copySource = constructCopySource(
+            requireSet(request.sourceBucket(), "sourceBucket"),
+            requireSet(request.sourceKey(), "sourceKey"),
+            request.sourceVersionId()
+        );
+        return request.toBuilder()
+                      .sourceBucket(null)
+                      .sourceKey(null)
+                      .sourceVersionId(null)
+                      .copySource(copySource)
+                      .build();
+    }
+
+    private static String constructCopySource(String sourceBucket, String sourceKey, String sourceVersionId) {
+        StringBuilder copySource = new StringBuilder();
+        copySource.append("/");
+        copySource.append(urlEncodeIgnoreSlashes(sourceBucket));
+        if (isArnFor(S3ResourceType.ACCESS_POINT, sourceBucket) || isArnFor(S3ResourceType.OUTPOST, sourceBucket)) {
+            copySource.append("/object");
+        }
```

Comment:
Yikes, didn't know we had to go this specific.

Reply:
Unfortunate, I agree. This logic was borrowed from .NET's [CopyObjectRequestMarshaller](https://github.com/aws/aws-sdk-net/blob/969f2c5f26e593fbb1c622eb2a9b16806b125e54/sdk/src/Services/S3/Custom/Model/Internal/MarshallTransformations/CopyObjectRequestMarshaller.cs#L159-L161).